From b0190320c3c1c521d8bdef292c55e1b33686a26f Mon Sep 17 00:00:00 2001
From: Alex Marginean <alexandru.marginean@freescale.com>
Date: Wed, 8 Nov 2017 10:44:54 +0200
Subject: [PATCH 075/706] enetc: disable SR-IOV when PF is removed

VF netdevices were left registered but unusable.

Signed-off-by: Alex Marginean <alexandru.marginean@freescale.com>
(cherry picked from commit 63695882a50e691b63daa9112a452db163a88dd1)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/ethernet/freescale/enetc/enetc.c | 53 +++++++++++---------
 1 file changed, 29 insertions(+), 24 deletions(-)

diff --git a/drivers/net/ethernet/freescale/enetc/enetc.c b/drivers/net/ethernet/freescale/enetc/enetc.c
index d7886e3ecdfe..a6718f4fcdd3 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc.c
+++ b/drivers/net/ethernet/freescale/enetc/enetc.c
@@ -1408,6 +1408,32 @@ static void enetc_free_msix(struct enetc_ndev_priv *priv)
 	kfree(priv->msix_entries);
 }
 
+#ifdef CONFIG_PCI_IOV
+static int enetc_sriov_configure(struct pci_dev *pdev, int num_vfs)
+{
+	struct enetc_si *si = pci_get_drvdata(pdev);
+	int err;
+
+	if (!num_vfs) {
+		dev_info(&pdev->dev, "SR-IOV stop\n");
+		pci_disable_sriov(pdev);
+	} else {
+		dev_info(&pdev->dev, "SR-IOV start, %d VFs\n", num_vfs);
+		err = pci_enable_sriov(pdev, num_vfs);
+		if (err) {
+			dev_err(&pdev->dev, "pci_enable_sriov err %d\n", err);
+			return err;
+		}
+	}
+
+	si->num_vfs = num_vfs;
+
+	return num_vfs;
+}
+#else
+#define enetc_sriov_configure(pdev, num_vfs)	(void)0
+#endif
+
 static int enetc_pci_probe(struct pci_dev *pdev,
 			   const struct pci_device_id *ent)
 {
@@ -1542,6 +1568,9 @@ static void enetc_pci_remove(struct pci_dev *pdev)
 
 	dev_info(&pdev->dev, "enetc_pci_remove()\n");
 
+	if (si->num_vfs)
+		enetc_sriov_configure(pdev, 0);
+
 	priv = netdev_priv(si->ndev);
 	unregister_netdev(si->ndev);
 
@@ -1557,30 +1586,6 @@ static void enetc_pci_remove(struct pci_dev *pdev)
 	pci_disable_device(pdev);
 }
 
-#ifdef CONFIG_PCI_IOV
-static int enetc_sriov_configure(struct pci_dev *pdev, int num_vfs)
-{
-	struct enetc_si *si = pci_get_drvdata(pdev);
-	int err;
-
-	if (!num_vfs) {
-		dev_info(&pdev->dev, "SR-IOV stop\n");
-		pci_disable_sriov(pdev);
-	} else {
-		dev_info(&pdev->dev, "SR-IOV start, %d VFs\n", num_vfs);
-		err = pci_enable_sriov(pdev, num_vfs);
-		if (err) {
-			dev_err(&pdev->dev, "pci_enable_sriov err %d\n", err);
-			return err;
-		}
-	}
-
-	si->num_vfs = num_vfs;
-
-	return num_vfs;
-}
-#endif
-
 static const struct pci_device_id enetc_id_table[] = {
 	{ PCI_DEVICE(PCI_VENDOR_ID_FREESCALE, ENETC_DEV_ID_PF) },
 	{ PCI_DEVICE(PCI_VENDOR_ID_FREESCALE, ENETC_DEV_ID_VF) },
-- 
2.17.1

